018-function-execution-model-refactor.md

```diff
@@ -0,0 +1,86 @@
+# Function Execution Model Rewrite
+- Implementation Owner: @PineappleIOnic
+- Start Date: 25-07-2021
+- Target Date: TBD
+
+## Summary
+
+[summary]: #summary
+
+<!-- Brief explanation of the proposed contribution. Write your answer below. -->
+This RFC contains a proposal for rewriting Cloud Functions in a way that will enable us to perform synchronous executions. This will introduce a new server into Appwrite specifically for dealing with executions called the "executor". The executor will deal with spinning up docker containers for runtimes that are required as well as updating containers with new code when needed. When API requests for functions are received the executor will forward them to the relevant runtime which will then handle it.
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+Cloud Functions currently do not support the ability to execute synchronous functions and will require a complete rethink to how we approach how we execute our functions in order for us to realize this goal.
+
+**What is the context or background in which this problem exists?**
+
+Cloud Functions are currently a bit limited, this new design idea will allow for more customizability and control of how functions will execute and run for users.
+
+**Once the proposal is implemented, how will the system change?**
+
+- A new process will be introduced to execute functions
+- All current runtimes will be updated to use a web server as it's core
+- The Functions worker will be stripped of all direct docker interactions and will instead call the executor
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+With this redesign of how cloud functions work the functions worker will no longer have to deal with spinning up docker containers but will instead use the executor server. Which will handle spinning up containers on behalf of the functions worker.
+
+The executor will have multiple tasks but the main one is to spin up runtimes when they are needed and to update their code whenever functions are updated. Runtimes will now be web servers based of their relevant languages and when code is added it will be added to the Runtime web server's router.
+
+The executor will also act like a router itself, when API Requests that are for functions are received it will be the executor's job to correlate them to the relevant runtimes and forward them for processing. When the Runtime's web server is done processing the request and returns a response the executor will return the response to whatever client requested it.
```

Comment:
I think here we need more technical description of how the executor it self will look, may be API signatures,
how worker/executor will communicate, some sample structures etc?

Reply:
I agree here